bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java

```diff
@@ -0,0 +1,546 @@
+     */
+    @Override
+    public void run() {
+        logger.info("Starting hub controller");
+        try {
+            Map<String, @Nullable Object> hubMsg;
+            String service;
+            hubReconnectSleep = HUB_RECONNECT_SLEEP_MIN;
+
+            // Main message loop listening for updates from the hub
+            while (!shouldStop()) {
+                if (!connect()) {
+                    continue;
+                }
+
+                if ((hubMsg = getHubMsg()) == null) {
+                    continue;
+                }
+
+                // Process message based on service type
+                if ((service = (String) hubMsg.get(HUB_TOKEN_SERVICE)) == null) {
+                    continue; // Ignore messages that don't have a service specified
+                }
+
+                if (service.equals(HUB_SERVICE_REPORT_ZONE_PROPERTIES)) {
+                    processMsgReportZoneProperties(hubMsg);
+                } else if (service.equals(HUB_SERVICE_ZONE_PROPERTIES_CHANGED)) {
+                    processMsgZonePropertiesChanged(hubMsg);
+                } else if (service.equals(HUB_SERVICE_LIST_ZONE)) {
+                    processMsgListZone(hubMsg);
+                } else if (service.equals(HUB_SERVICE_SYSTEM_INFO)) {
+                    processMsgSystemInfo(hubMsg);
+                }
+            }
+        } catch (Exception e) {
+            logger.error("Hub controller failed ({})", e.getMessage());
+        } finally {
+            synchronized (this) {
+                // If there are still pending requests we need to cancel them
+                for (AdorneHubController.ZoneIDFutureRecord stateCommand : stateCommands) {
+                    stateCommand.future.cancel(false);
+                }
+                for (CompletableFuture<List<Integer>> zoneCommand : zoneCommands) {
+                    zoneCommand.cancel(false);
+                }
+                for (CompletableFuture<String> macAddressCommand : macAddressCommands) {
+                    macAddressCommand.cancel(false);
+                }
+                CompletableFuture<@Nullable Void> hubControllerConnected = this.hubControllerConnected;
+                hubControllerConnected.cancel(false);
+            }
+            disconnect();
+            logger.info("Exiting hub controller");
+        }
+    }
+
+    private synchronized boolean shouldStop() {
+        boolean timeoutExceeded = stopTimestamp > 0 && System.currentTimeMillis() > stopTimestamp;
+        boolean commandsServed = stopWhenCommandsServed.get() && stateCommands.isEmpty() && zoneCommands.isEmpty()
+                && macAddressCommands.isEmpty();
+
+        return (stop.get() || timeoutExceeded || commandsServed);
+    }
+
+    private boolean connect() {
+        try {
+            synchronized (this) {
+                if (hubSocket == null) {
+                    Socket hubSocket = this.hubSocket = new Socket(hubHost, hubPort);
```

Comment:
Please  split this in two statements (also applies below)

Reply:
Done